shared-libs/transitions/src/transitions/index.js

```diff
@@ -96,60 +52,79 @@ const processChange = (change, callback) => {
         if (change.doc.transitions) {
           delete change.doc.transitions;
         }
-        module.exports.applyTransitions(change, () => {
-          processed++;
-          metadata
-            .update(change.seq)
-            .then(() => callback())
-            .catch(err => {
-              callback(err);
-            });
-        });
+        applyTransitions(change, callback);
       });
     })
     .catch(err => {
       logger.error('transitions: fetch failed for %s : %o', change.id, err);
-      return callback();
+      return callback(err);
     });
 };
 
-const deleteReadDocs = change => {
-  // we don't know if the deleted doc was a report or a message so
-  // attempt to delete both
-  const possibleReadDocIds = ['report', 'message'].map(
-    type => `read:${type}:${change.id}`
-  );
+// given a collection of docs this function will:
+// - deep clone the docs, to keep the original version of them safe
+// - add _id properties to the docs that are missing them
+// - hydrate the docs and generate info docs
+// - run loaded transitions over all docs
+// - returns docs, either updated or their original version
+const processDocs = docs => {
+  if (!transitions.length) {
+    return Promise.resolve(docs);
+  }
 
-  return db.allDbs().then(dbs => {
-    const userDbs = dbs.filter(db => db.indexOf('medic-user-') === 0);
-    return userDbs.reduce((p, userDb) => {
-      return p.then(() => {
-        const metaDb = db.get(userDb);
-        return metaDb.allDocs({ keys: possibleReadDocIds }).then(results => {
-          const row = results.rows.find(row => !row.error);
-          if (!row) {
-            return;
-          }
+  let changes;
+  const clonedDocs = JSON.parse(JSON.stringify(docs)),
```

Comment:
Yeah every time I check this is what it comes back to. Last time I checked was a couple of months ago.

Reply:
I guess it depends on what you're cloning. 
As a result of Stefan's comment, I tried a couple of _reliable_ alternatives : http://jsben.ch/SJghM
- [lodash cloneDeep](https://lodash.com/docs/4.17.11#cloneDeep) 
- [angular copy](https://docs.angularjs.org/api/ng/function/angular.copy)

Do you guys think lodash.cloneDeep would be worth including? It can be required as a standalone.

Reply:
In my Chromium, JSON is faster than lodash but angular is way faster. Do you get lodash as fastest?

![screenshot from 2019-03-08 09-32-34](https://user-images.githubusercontent.com/2064938/53987176-2349b680-4185-11e9-9556-33aa76dcbb42.png)

Reply:
_slightly_ faster

![image](https://user-images.githubusercontent.com/35681649/53987481-ca006980-4129-11e9-9431-e770d29d1cb2.png)

Reply:
Mine's consistently slower. Strange. Different browser versions maybe?

I wonder how angular copy is doing it?

I don't think it's worth switching to lodash but it may be worth ripping out the angular version and using that throughout in a separate PR.

Reply:
I was planning on looking at the Angular version tomorrow, it got me curious. 

The lodash seems to be the complete solution to deep copy and tries to treat all possible cases, while angular's solution is clearly stated to be riddled with caveats, albeit faster.